Ambient sounds animations and civ dialogs

UAC-MaxxLite opened the pull request:
## What type of PR is this?
1. [ ] Bug
2. [x] Change
3. [x] Enhancement
4. [ ] Miscellaneous

### What have you changed and why?
Information:
 - revamps the random civ radio system :
    * now spawns a radio item that can be destroyed to remove the music.
    
 - added animations and sounds to petros.
 - added an option to interact with civs to get a rescue mission. ( also has a chance to reveal an enemy location if option to hide location is enabled ).
 
 - moved findIfNearAndEnemy to its own function and split it to two different ones.
      * findIfNearAndEnemyZones - finds all near enemy zones in a list from closest.
      * findIfNearAndEnemy - finds the nearest enemy zone.

### Please specify which Issue this PR Resolves (If Applicable).
"This PR closes #XXXX!"

### Please verify the following.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [ ] No
2. [x] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps:
check the civs with high pop per city to see what happens.
********************************************************
Notes:

UAC-MaxxLite committed:
updating sound scape 1/2

UAC-MaxxLite committed:
Updating Sound Scape 2/2?

UAC-MaxxLite committed:
more features!

added animations
civ interaction
fixed issues
made find findIfNearAndHostile into its own fnc
civs have a chance to reveal zones if hide zones is set to true

and some more i cant remember

UAC-MaxxLite committed:
spot and error fixes

UAC-MaxxLite committed:
Some fixes and split a function into 2

UAC-MaxxLite committed:
finishing touches

Silence-112 commented on A3A/addons/patcom/sounds/Civilian/Ambient/Sneeze.ogg:
Is this meant to be entirely in the left ear?

UAC-MaxxLite replied:
nope. seems like there is an issue with the recording. will look into it

UAC-MaxxLite replied:
nope. seems like there is an issue with the recording. will look into it

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivilian.sqf:
Ideally rename to `fn_dialogCiv.sqf` to maintain consistency with `fn_dialogCivFinished.sqf`

UAC-MaxxLite replied:
good idea. done

Silence-112 commented on A3A/addons/core/functions/Missions/fn_findIfNearAndHostile.sqf:
```diff
@@ -0,0 +1,29 @@
+    Description:
+        Finds the closest enemy town to a selected refrance unit/object
+    
+    Params:
+        _Markers <Array> <Default: "">
+        _unit <unit/object> <Default: petros>
+    
+    Usage:
+        [_Markers, _unit] call A3A_fnc_findIfNearAndHostile;
+    
+    Return:
+        The closest enemy controlled marker
+*/
+
+params [  
+  ["_Markers", ""],  
+  ["_unit", petros]  
+]; 
+
+private _referencePos = getPosWorld _unit; 
+private _nearestMarkers = [allMapMarkers, _referencePos] call BIS_fnc_nearestPosition; 
+_Markers = _Markers select {(getMarkerPos _x distance2D getMarkerPos _nearestMarkers < (distanceMission * 2)) && (sidesX getVariable [_x,sideUnknown] != teamPlayer)}; 
+_Markers = [_Markers,[],{_referencePos distanceSqr getMarkerPos _x},"ASCEND"] call BIS_fnc_sortBy; 
+_nearestMarker = _Markers select 0; 
+_nearestMarker
```
Corrected default values and variable case scheme
```suggestion
    Params:
        _markers <Array> <Default: []>
        _unit <Unit/Object> <Default: ObjNull>
    
    Usage:
        [_markers, _unit] call A3A_fnc_findIfNearAndHostile;
    
    Return:
        The closest enemy controlled marker
*/

params [  
  ["_markers", []],  
  ["_unit", ObjNull]  
]; 

private _referencePos = getPosWorld _unit; 
private _nearestMarkers = [allMapMarkers, _referencePos] call BIS_fnc_nearestPosition; 
_markers = _markers select {(getMarkerPos _x distance2D getMarkerPos _nearestMarkers < (distanceMission * 2)) && (sidesX getVariable [_x,sideUnknown] != teamPlayer)}; 
_markers = [_markers,[],{_referencePos distanceSqr getMarkerPos _x},"ASCEND"] call BIS_fnc_sortBy; 
_nearestMarker = _markers select 0; 
_nearestMarker
```

UAC-MaxxLite replied:
done

Silence-112 commented on A3A/addons/core/functions/Missions/fn_findIfNearAndHostileZones.sqf:
Same feedback as above and in addition couldn't you just merge `findIfNearAndHostile(Zones).sqf` into one function and change the return value accordingly?

UAC-MaxxLite replied:
yeah i didnt think of that to be honest

UAC-MaxxLite replied:
done

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_civilianInitEH.sqf:
```diff
@@ -27,7 +27,7 @@ _unit setSkill 0;
 _unit disableAI "AUTOTARGET";
 //Stops civilians from shouting out commands.
-[_unit, createHashMapFromArray [["face", selectRandom (A3A_faction_civ get "faces")], ["speaker", "NoVoice"]]] call A3A_fnc_setIdentity;
+[_unit, selectRandom (A3A_faction_civ get "faces"), "NoVoice"] call A3A_fnc_setIdentity;
```
This change will cause issues since setIdentity requires a hashmap post #346

UAC-MaxxLite replied:
guess this was a remanent from original file as i dont remember adding that

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_unitAmbient.sqf:
```diff
@@ -0,0 +1,62 @@
+        _return <_unit>
+*/
+
+params [["_musicSource", petros]];
```
Good practice to prevent function misuse
```suggestion
params [["_musicSource", ObjNull]];

if (_musicSource isEqualTo ObjNull) exitWith {nil};
```

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_unitAmbient.sqf:
```diff
@@ -0,0 +1,62 @@
+ 
+		private _animation = _sound # 1; 
+ 
+		[_musicSource, _sound # 0] remoteExec ["say3D", [1, _musicSource], true]; 
```
`[1, _musicSource]`
Does this just play the sound where _musicSource is local, which would likely be the server?

Silence-112 replied:
[The wiki](https://community.bistudio.com/wiki/Multiplayer_Scripting#Machine_network_ID) states that 1 should not be used as it isn't implemented

0 here should suffice, playing the sound globally as intended

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivilian.sqf:
```diff
@@ -0,0 +1,46 @@
+
+#include "..\..\script_component.hpp"
+private _debugStatus = format ["[Civ Dialog] | Civ Hold Action | caller: %1 Civ Unit: %2", _caller, _civUnit];
+params ["_civUnit", nil];
```
Param defaults should reflect the type they represent
```suggestion
params ["_civUnit", ObjNull];
```

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivilian.sqf:
```diff
@@ -0,0 +1,46 @@
+
+
+#include "..\..\script_component.hpp"
+private _debugStatus = format ["[Civ Dialog] | Civ Hold Action | caller: %1 Civ Unit: %2", _caller, _civUnit];
```
You have a variable here that never gets used?

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+        on success selects a mission and assigns it to the closest enemy town.
+    
+    Params:
+        _civUnit <unit> <Default: nil>
```
This default isn't factual since no params are actually provided

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+
+#include "..\..\script_component.hpp"
+
+private _civilian = _this select 0;
```
Any particular reason it was decided not to use `params []` here?

UAC-MaxxLite replied:
didnt think about it at the time mb

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+_possibleMarkers = _possibleMarkers select 1;
+
+private _dialogResult = "";
+private _civrequestedMission = "";
```
```suggestion
private _civRequestedMission = "";
```

Change all variable references accordingly

UAC-MaxxLite replied:
done

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+
+        [[_possibleMarkers],_civrequestedMission] remoteExec ["A3A_fnc_scheduler",2];
+        if (hideEnemyMarkers) then {
+            if (10 >= random(100)) then {
```
```suggestion
            if (10 >= (random 100)) then {
```

UAC-MaxxLite replied:
fixed

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+    _dialogResult = "Failure";
+
+    if (hideEnemyMarkers) then {
+        if (5 >= random(100)) then {
```
```suggestion
        if (5 >= (random 100)) then {
```

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,131 @@
+                localize "STR_antistasi_actions_talk_with_civ_fail_zone_reveal2"
+            ]);
+
+            [1, "A civilian as revealed a zone"] call A3U_fnc_revealRandomZones;
```
```suggestion
            [1, "A civilian has revealed a zone"] call A3U_fnc_revealRandomZones;
```

Silence-112 reviewed (commented):
Will need very extensive testing when it gets to a good enough state

UAC-MaxxLite committed:
Update A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf

Co-authored-by: Silence <78276788+SilenceIsFatto@users.noreply.github.com>

UAC-MaxxLite committed:
Update A3A/addons/core/functions/Missions/fn_findIfNearAndHostile.sqf

Co-authored-by: Silence <78276788+SilenceIsFatto@users.noreply.github.com>

UAC-MaxxLite committed:
recommended fixes

UAC-MaxxLite commented:
NOTE: current changes untested due to time constraints and doing it from my phone....

UAC-MaxxLite committed:
findIfNearAndHostile fix

UAC-MaxxLite committed:
fixes with hold action

Silence-112 commented on A3A/addons/patcom/functions/Civilian/fn_dialogCivFinished.sqf:
```diff
@@ -0,0 +1,127 @@
+    _dialogResult = "Failure";
+
+    if (hideEnemyMarkers) then {
+        if (5 >= (random 100)) then {
```
5 seems a bit low, won't this practically always be true

Silence-112 reviewed (commented):
Looks better, will test at some point
Also i'm not sure if you know but some of the review comments are hidden, the button to unhide them should look like the below
![image](https://github.com/user-attachments/assets/0bebe5aa-ace0-4ce6-bfcc-27f497666f8b)

Silence-112 committed:
Merge branch 'unstable' into ambient-sounds

Silence-112 committed:
Fixes and improvements

Feedback:
Params that have a default value MUST be `params [["_variable", _TYPE_]]` rather than `params ["_variable", _TYPE_]` (Makes the function think that `_TYPE_` is also a variable)

Make sure any param default values are the type that the variable is meant to hold. E.g, _unit default should be ObjNull rather than nil

Rather than repeating code (see `_civUnit globalChat ...` in dialogCivFinished), store the parameter for it as a variable and run the code once at the end

Don't use sleep when the entire function isn't, or doesn't have to be, scheduled. You can however use CBA_fnc_waitAndExecute, which only schedules the code needed. See lines ~80-100 in dialogCivFinished

Make sure the function headers are actually accurate and the listed params contain only what is needed. E.g `_civUnit <OBJECT> <Default: ObjNull>`

Silence-112 committed:
Civ intel rework, radio item changes

Silence-112 committed:
Small interaction bugfix

Silence-112 committed:
Yet another small interaction bugfix

UAC-MaxxLite committed:
animations to go globaly

Silence-112 committed:
unitAmbient cleanup

Silence-112 committed:
Merge branch 'unstable' into ambient-sounds

Silence-112 committed:
Added dialogue parameter, improved networking

Silence-112 committed:
Merge branch 'unstable' into ambient-sounds
